File: Sources/UsageBarLocalAPI/LocalUsageAPIProtocol.swift
import Foundation

struct LocalUsageAPIHTTPResponseBuilder {
    func build(status: LocalUsageAPIHTTPStatus, body: Data = Data(), date: Date) -> Data {
        let responseBody = status == .ok ? body : Data()
        var lines = [
            "HTTP/1.1 \(status.rawValue) \(status.reasonPhrase)",
            "Date: \(Self.httpDate(date))"
        ]
        if status == .ok {
            lines.append("Content-Type: application/json; charset=utf-8")
        }
        lines.append("Content-Length: \(responseBody.count)")
        lines.append("Connection: close")
        lines.append("Cache-Control: no-store")
        lines.append("X-Content-Type-Options: nosniff")
        if status == .methodNotAllowed {
            lines.append("Allow: GET")
        }
        if status == .tooManyRequests {
            lines.append("Retry-After: 1")
        }
        lines.append("")
        lines.append("")

        var result = Data(lines.joined(separator: "\r\n").utf8)
        result.append(responseBody)
        return result
    }

    static func httpDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss 'GMT'"
        return formatter.string(from: date)
    }
}

final class LocalUsageAPITokenBucket {
    private let lock = NSLock()
    private let rate: Double
    private let capacity: Double
    private var tokens: Double
    private var lastMonotonicTime: TimeInterval

    init(ratePerSecond: Double = 4, burst: Int = 8, initialMonotonicTime: TimeInterval) {
        precondition(ratePerSecond > 0)
        precondition(burst > 0)
        rate = ratePerSecond
        capacity = Double(burst)
        tokens = Double(burst)
        lastMonotonicTime = initialMonotonicTime
    }

    func admit(at monotonicTime: TimeInterval) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        let elapsed = max(0, monotonicTime - lastMonotonicTime)
        tokens = min(capacity, tokens + elapsed * rate)
        lastMonotonicTime = max(lastMonotonicTime, monotonicTime)
        guard tokens >= 1 else { return false }
        tokens -= 1
        return true
    }
}

final class LocalUsageAPIActiveConnectionGate {
    private let lock = NSLock()
    private let maximum: Int
    private var active = 0

    init(maximum: Int = 4) {
        precondition(maximum > 0)
        self.maximum = maximum
    }

    func acquire() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        guard active < maximum else { return false }
        active += 1
        return true
    }

    func release() {
        lock.lock()
        defer { lock.unlock() }
        precondition(active > 0)
        active -= 1
    }

    var activeCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return active
    }
}

protocol LocalUsageAPIDeadlineToken: AnyObject {
    func cancel()
}

protocol LocalUsageAPIDeadlineScheduling {
    func schedule(after delay: TimeInterval, action: @escaping () -> Void) -> LocalUsageAPIDeadlineToken
}

final class DispatchLocalUsageAPIDeadlineScheduler: LocalUsageAPIDeadlineScheduling {
    private let queue: DispatchQueue

    init(queue: DispatchQueue) {
        self.queue = queue
    }

    func schedule(after delay: TimeInterval, action: @escaping () -> Void) -> LocalUsageAPIDeadlineToken {
        let item = DispatchWorkItem(block: action)
        queue.asyncAfter(deadline: .now() + delay, execute: item)
        return DispatchDeadlineToken(item: item)
    }
}

private final class DispatchDeadlineToken: LocalUsageAPIDeadlineToken {
    private let item: DispatchWorkItem

    init(item: DispatchWorkItem) {
        self.item = item
    }

    func cancel() {
        item.cancel()
    }
}

final class LocalUsageAPIAbsoluteDeadline {
    private let delay: TimeInterval
    private let scheduler: LocalUsageAPIDeadlineScheduling
    private var token: LocalUsageAPIDeadlineToken?

    init(delay: TimeInterval, scheduler: LocalUsageAPIDeadlineScheduling) {
        self.delay = delay
        self.scheduler = scheduler
    }

    func start(action: @escaping () -> Void) {
        guard token == nil else { return }
        token = scheduler.schedule(after: delay, action: action)
    }

    func cancel() {
        token?.cancel()
        token = nil
    }
}

final class LocalUsageAPIRequestProcessor {
    typealias SnapshotProvider = (Date, @escaping (Result<Data, Error>) -> Void) -> Void

    static let maximumResponseBodyBytes = 16 * 1_024

    private let parser: StrictHTTPRequestParser
    private let responseBuilder: LocalUsageAPIHTTPResponseBuilder
    private let wallNow: () -> Date
    private let snapshotProvider: SnapshotProvider

    init(
        expectedHost: String,
        wallNow: @escaping () -> Date,
        snapshotProvider: @escaping SnapshotProvider
    ) {
        parser = StrictHTTPRequestParser(expectedHost: expectedHost)
        responseBuilder = LocalUsageAPIHTTPResponseBuilder()
        self.wallNow = wallNow
        self.snapshotProvider = snapshotProvider
    }

    @discardableResult
    func process(_ request: Data, completion: @escaping (Data) -> Void) -> Bool {
        switch parser.parse(request) {
        case .incomplete:
            return false
        case let .rejected(status):
            completion(responseBuilder.build(status: status, date: wallNow()))
            return true
        case .accepted:
            let observedAt = wallNow()
            snapshotProvider(observedAt) { result in
                let response: Data
                switch result {
                case let .success(body) where body.count <= Self.maximumResponseBodyBytes:
                    response = self.responseBuilder.build(status: .ok, body: body, date: observedAt)
                case .success, .failure:
                    response = self.responseBuilder.build(status: .serviceUnavailable, date: observedAt)
                }
                completion(response)
            }
            return true
        }
    }

    func response(status: LocalUsageAPIHTTPStatus) -> Data {
        responseBuilder.build(status: status, date: wallNow())
    }
}

File: windows/src/UsageBar.Windows.Infrastructure/LocalApi/LocalUsageApiProtocol.cs
using System.Globalization;
using System.Text;

namespace UsageBar.Windows.Infrastructure.LocalApi;

internal enum LocalUsageApiHttpStatus
{
    Ok = 200,
    BadRequest = 400,
    NotFound = 404,
    MethodNotAllowed = 405,
    ContentTooLarge = 413,
    UriTooLong = 414,
    TooManyRequests = 429,
    RequestHeaderFieldsTooLarge = 431,
    ServiceUnavailable = 503,
    HttpVersionNotSupported = 505
}

public delegate Task<byte[]> WindowsLocalUsageApiSnapshotProvider(
    DateTimeOffset observedAt,
    CancellationToken cancellationToken);

internal sealed class LocalUsageApiHttpResponseBuilder
{
    public byte[] Build(
        LocalUsageApiHttpStatus status,
        ReadOnlyMemory<byte> body,
        DateTimeOffset date)
    {
        var responseBody = status == LocalUsageApiHttpStatus.Ok ? body : ReadOnlyMemory<byte>.Empty;
        var lines = new List<string>
        {
            $"HTTP/1.1 {(int)status} {ReasonPhrase(status)}",
            $"Date: {HttpDate(date)}"
        };

        if (status == LocalUsageApiHttpStatus.Ok)
        {
            lines.Add("Content-Type: application/json; charset=utf-8");
        }

        lines.Add($"Content-Length: {responseBody.Length.ToString(CultureInfo.InvariantCulture)}");
        lines.Add("Connection: close");
        lines.Add("Cache-Control: no-store");
        lines.Add("X-Content-Type-Options: nosniff");

        if (status == LocalUsageApiHttpStatus.MethodNotAllowed)
        {
            lines.Add("Allow: GET");
        }

        if (status == LocalUsageApiHttpStatus.TooManyRequests)
        {
            lines.Add("Retry-After: 1");
        }

        lines.Add(string.Empty);
        lines.Add(string.Empty);

        var header = Encoding.ASCII.GetBytes(string.Join("\r\n", lines));
        var response = new byte[header.Length + responseBody.Length];
        header.CopyTo(response, 0);
        responseBody.Span.CopyTo(response.AsSpan(header.Length));
        return response;
    }

    public byte[] Build(LocalUsageApiHttpStatus status, DateTimeOffset date) =>
        Build(status, ReadOnlyMemory<byte>.Empty, date);

    internal static string HttpDate(DateTimeOffset date) =>
        date.UtcDateTime.ToString("R", CultureInfo.InvariantCulture);

    private static string ReasonPhrase(LocalUsageApiHttpStatus status) => status switch
    {
        LocalUsageApiHttpStatus.Ok => "OK",
        LocalUsageApiHttpStatus.BadRequest => "Bad Request",
        LocalUsageApiHttpStatus.NotFound => "Not Found",
        LocalUsageApiHttpStatus.MethodNotAllowed => "Method Not Allowed",
        LocalUsageApiHttpStatus.ContentTooLarge => "Content Too Large",
        LocalUsageApiHttpStatus.UriTooLong => "URI Too Long",
        LocalUsageApiHttpStatus.TooManyRequests => "Too Many Requests",
        LocalUsageApiHttpStatus.RequestHeaderFieldsTooLarge => "Request Header Fields Too Large",
        LocalUsageApiHttpStatus.ServiceUnavailable => "Service Unavailable",
        LocalUsageApiHttpStatus.HttpVersionNotSupported => "HTTP Version Not Supported",
        _ => throw new ArgumentOutOfRangeException(nameof(status))
    };
}

internal sealed class LocalUsageApiTokenBucket
{
    private readonly object _lock = new();
    private readonly double _rate;
    private readonly double _capacity;
    private double _tokens;
    private double _lastMonotonicTime;

    public LocalUsageApiTokenBucket(
        double initialMonotonicTime,
        double ratePerSecond = 4,
        int burst = 8)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(initialMonotonicTime);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(ratePerSecond);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(burst);
        _rate = ratePerSecond;
        _capacity = burst;
        _tokens = burst;
        _lastMonotonicTime = initialMonotonicTime;
    }

    public bool Admit(double monotonicTime)
    {
        lock (_lock)
        {
            var elapsed = Math.Max(0, monotonicTime - _lastMonotonicTime);
            _tokens = Math.Min(_capacity, _tokens + elapsed * _rate);
            _lastMonotonicTime = Math.Max(_lastMonotonicTime, monotonicTime);
            if (_tokens < 1)
            {
                return false;
            }

            _tokens -= 1;
            return true;
        }
    }
}

internal sealed class LocalUsageApiActiveConnectionGate
{
    private readonly object _lock = new();
    private readonly int _maximum;
    private int _active;

    public LocalUsageApiActiveConnectionGate(int maximum = 4)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(maximum);
        _maximum = maximum;
    }

    public bool TryAcquire()
    {
        lock (_lock)
        {
            if (_active >= _maximum)
            {
                return false;
            }

            _active += 1;
            return true;
        }
    }

    public void Release()
    {
        lock (_lock)
        {
            if (_active <= 0)
            {
                throw new InvalidOperationException("No active local API connection to release.");
            }

            _active -= 1;
        }
    }

    public int ActiveCount
    {
        get
        {
            lock (_lock)
            {
                return _active;
            }
        }
    }
}

internal interface ILocalUsageApiDeadline : IDisposable
{
    void Cancel();
}

internal interface ILocalUsageApiDeadlineScheduler
{
    ILocalUsageApiDeadline Schedule(TimeSpan delay, Action action);
}

internal sealed class TimerLocalUsageApiDeadlineScheduler : ILocalUsageApiDeadlineScheduler
{
    public ILocalUsageApiDeadline Schedule(TimeSpan delay, Action action) =>
        new TimerLocalUsageApiDeadline(delay, action);

    private sealed class TimerLocalUsageApiDeadline : ILocalUsageApiDeadline
    {
        private Timer? _timer;

        public TimerLocalUsageApiDeadline(TimeSpan delay, Action action)
        {
            ArgumentNullException.ThrowIfNull(action);
            _timer = new Timer(_ => action(), null, delay, Timeout.InfiniteTimeSpan);
        }

        public void Cancel() => Dispose();

        public void Dispose() => Interlocked.Exchange(ref _timer, null)?.Dispose();
    }
}

internal sealed class LocalUsageApiAbsoluteDeadline : IDisposable
{
    private readonly TimeSpan _delay;
    private readonly ILocalUsageApiDeadlineScheduler _scheduler;
    private ILocalUsageApiDeadline? _scheduled;

    public LocalUsageApiAbsoluteDeadline(
        TimeSpan delay,
        ILocalUsageApiDeadlineScheduler scheduler)
    {
        _delay = delay;
        _scheduler = scheduler;
    }

    public void Start(Action action)
    {
        _scheduled ??= _scheduler.Schedule(_delay, action);
    }

    public void Cancel()
    {
        Interlocked.Exchange(ref _scheduled, null)?.Cancel();
    }

    public void Dispose() => Cancel();
}

internal readonly record struct LocalUsageApiProcessResult(bool IsComplete, byte[]? Response);

internal sealed class LocalUsageApiRequestProcessor
{
    public const int MaximumResponseBodyBytes = 16 * 1024;

    private readonly StrictHttpRequestParser _parser;
    private readonly LocalUsageApiHttpResponseBuilder _responseBuilder = new();
    private readonly Func<DateTimeOffset> _wallNow;
    private readonly WindowsLocalUsageApiSnapshotProvider _snapshotProvider;

    public LocalUsageApiRequestProcessor(
        string expectedHost,
        Func<DateTimeOffset> wallNow,
        WindowsLocalUsageApiSnapshotProvider snapshotProvider)
    {
        _parser = new StrictHttpRequestParser(expectedHost);
        _wallNow = wallNow;
        _snapshotProvider = snapshotProvider;
    }

    public async Task<LocalUsageApiProcessResult> ProcessAsync(
        ReadOnlyMemory<byte> request,
        CancellationToken cancellationToken = default)
    {
        var parsed = _parser.Parse(request.Span);
        if (parsed.Kind == StrictHttpRequestParseResultKind.Incomplete)
        {
            return new LocalUsageApiProcessResult(false, null);
        }

        if (parsed.Kind == StrictHttpRequestParseResultKind.Rejected)
        {
            return new LocalUsageApiProcessResult(
                true,
                Response(parsed.Status!.Value));
        }

        var observedAt = _wallNow();
        try
        {
            var body = await _snapshotProvider(observedAt, cancellationToken).ConfigureAwait(false);
            if (body.Length > MaximumResponseBodyBytes)
            {
                return new LocalUsageApiProcessResult(
                    true,
                    _responseBuilder.Build(LocalUsageApiHttpStatus.ServiceUnavailable, observedAt));
            }

            return new LocalUsageApiProcessResult(
                true,
                _responseBuilder.Build(LocalUsageApiHttpStatus.Ok, body, observedAt));
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw;
        }
        catch (Exception)
        {
            return new LocalUsageApiProcessResult(
                true,
                _responseBuilder.Build(LocalUsageApiHttpStatus.ServiceUnavailable, observedAt));
        }
    }

    public byte[] Response(LocalUsageApiHttpStatus status) =>
        _responseBuilder.Build(status, _wallNow());
}
